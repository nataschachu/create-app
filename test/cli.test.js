/* eslint-disable @typescript-eslint/no-var-requires, global-require */
const fs = require('fs');
const path = require('path');

const cmd = require('./runner');

const defaultInputs = [
    {
        regexpToWait: /Название вашего проекта/,
        textToInput: `template-app${cmd.ENTER}`,
    },
    {
        regexpToWait: /Хотите изменить список стандартных функций проекта/,
        textToInput: `n${cmd.ENTER}`,
    },
    {
        regexpToWait: /Выполнить установку зависимостей \(npm install\)/,
        textToInput: `n${cmd.ENTER}`,
    },
];

async function runCli(inputs = [], ...args) {
    const cliProcess = cmd.create();

    return cliProcess.execute(
        [
            '--experimental-import-meta-resolve',
            './src/cliWizard.mjs',
            '--runDir',
            path.dirname(__dirname),
            '--templatePackage',
            '@sberdevices/create-canvas-app',
            ...args,
        ], // args
        inputs,
        {
            env: {
                ...process.env,
                INSIDE_CLI_TEST_RUNNER: true,
            },
        },
    );
}

describe('cli', () => {
    afterEach(() => {
        fs.rmdirSync('./template-app/', { recursive: true, force: true });
    });

    it('default case of template', async () => {
        const response = await runCli(defaultInputs);

        console.log(response);
        expect(response).toMatchSnapshot();

        const packageJsonExists = fs.existsSync('./template-app/package.json');
        expect(packageJsonExists).toBe(true);
    });

    it('template without feature toggles', async () => {
        const inputs = [defaultInputs[0], defaultInputs[2]];

        const response = await runCli(inputs, '--configPath', `${__dirname}/__mock__/emptyConfig.mjs`);

        console.log(response);
        expect(response).toMatchSnapshot();

        const packageJsonExists = fs.existsSync('./template-app/package.json');
        expect(packageJsonExists).toBe(true);
    });

    it('mustache applied to PageIndex', async () => {
        await runCli(defaultInputs);

        const filePathToCheck = './template-app/src/components/PageIndex.tsx';
        const fileExists = fs.existsSync(filePathToCheck);

        expect(fileExists).toBe(true);

        const fileContent = fs.readFileSync(filePathToCheck).toString();

        expect(fileContent).toMatchSnapshot();
    });
});
