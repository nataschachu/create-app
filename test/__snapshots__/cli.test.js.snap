// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli default case of template 1`] = `
"[32m __   __   ___  __   __   ___         __   ___  __    /  __   __   ___      ___  ___          __   __  [39m
[32m/__\` |__) |__  |__) |  \\\\ |__  \\\\  / | /  \` |__  /__\`  /  /  \` |__) |__   /\\\\   |  |__  __  /\\\\  |__) |__) [39m
[32m.__/ |__) |___ |  \\\\ |__/ |___  \\\\/  | \\\\__, |___ .__/ /   \\\\__, |  \\\\ |___ /~~\\\\  |  |___    /~~\\\\ |    |    [39m
[32m                                                                                                       [39m
‚ÑπÔ∏è  –ó–∞–≥—Ä—É–∑–∫–∞ —à–∞–±–ª–æ–Ω–∞–û–ø–∏—Å–∞–Ω–∏–µ —à–∞–±–ª–æ–Ω–∞: @sberdevices/create-canvas-app ‚Äì –≤–∏–∑–∞—Ä–¥ –¥–ª—è –±—ã—Å—Ç—Ä–æ–≥–æ —Å–æ–∑–¥–∞–Ω–∏—è canvas-app c –Ω—É–ª—è –Ω–∞ –±–∞–∑–µ nextjs

? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) [42D[42C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) t[43D[43C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) te[44D[44C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) tem[45D[45C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) temp[46D[46C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) templ[47D[47C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) templa[48D[48C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) templat[49D[49C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) template[50D[50C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) template-[51D[51C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) template-a[52D[52C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) template-ap[53D[53C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) template-app[54D[54C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: template-app[39D[39C
? –•–æ—Ç–∏—Ç–µ –∏–∑–º–µ–Ω–∏—Ç—å —Å–ø–∏—Å–æ–∫ —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã—Ö —Ñ—É–Ω–∫—Ü–∏–π –ø—Ä–æ–µ–∫—Ç–∞? (y/N) [60D[60C[2K[G? –•–æ—Ç–∏—Ç–µ –∏–∑–º–µ–Ω–∏—Ç—å —Å–ø–∏—Å–æ–∫ —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã—Ö —Ñ—É–Ω–∫—Ü–∏–π –ø—Ä–æ–µ–∫—Ç–∞? (y/N) n[61D[61C[2K[G? –•–æ—Ç–∏—Ç–µ –∏–∑–º–µ–Ω–∏—Ç—å —Å–ø–∏—Å–æ–∫ —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã—Ö —Ñ—É–Ω–∫—Ü–∏–π –ø—Ä–æ–µ–∫—Ç–∞? No[56D[56C
? –í—ã–ø–æ–ª–Ω–∏—Ç—å —É—Å—Ç–∞–Ω–æ–≤–∫—É –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–µ–π (npm install)? (Y/n) [56D[56C[2K[G? –í—ã–ø–æ–ª–Ω–∏—Ç—å —É—Å—Ç–∞–Ω–æ–≤–∫—É –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–µ–π (npm install)? (Y/n) n[57D[57C[2K[G? –í—ã–ø–æ–ª–Ω–∏—Ç—å —É—Å—Ç–∞–Ω–æ–≤–∫—É –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–µ–π (npm install)? No[52D[52C

‚úÖ –ì–æ—Ç–æ–≤–æ!

"
`;

exports[`cli mustache applied to PageIndex 1`] = `
"import { Header } from '@sberdevices/plasma-ui';

import { useBasePath } from '../utils';
import { useCharacter } from '../utils/character';

import { LinkList } from './LinkList';

export const PageIndex = ({ component }: { component: React.ReactNode }) => {
    const basePath = useBasePath();
    const character = useCharacter();
    const characterElement = \`–ü–µ—Ä—Å–æ–Ω–∞–∂: \${character}\`;

    return (
        <>
            <Header title=\\"–î–æ–±—Ä–æ –ø–æ–∂–∞–ª–æ–≤–∞—Ç—å –≤ SberDevices!\\" />
            {component}
            { characterElement }
            {}
            <LinkList
                links={[
                    { href: '/SSR', content: 'Server-side rendering' },
                    { href: \`\${basePath}/SSG\`, content: 'Static Generation / Incremental Static Regeneration' },
                    { href: \`\${basePath}/prefetch\`, content: '–ü—Ä–µ—Ñ–µ—Ç—á —Å–ª–µ–¥—É—é—â–µ–π —Å—Ç—Ä–∞–Ω–∏—Ü—ã' },
                    { href: \`\${basePath}/image-optimization\`, content: '–û–ø—Ç–∏–º–∏–∑–∞—Ü–∏—è –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏–π' },
                ]}
            />
        </>
    );
};
"
`;

exports[`cli template without feature toggles 1`] = `
"[32m __   __   ___  __   __   ___         __   ___  __    /  __   __   ___      ___  ___          __   __  [39m
[32m/__\` |__) |__  |__) |  \\\\ |__  \\\\  / | /  \` |__  /__\`  /  /  \` |__) |__   /\\\\   |  |__  __  /\\\\  |__) |__) [39m
[32m.__/ |__) |___ |  \\\\ |__/ |___  \\\\/  | \\\\__, |___ .__/ /   \\\\__, |  \\\\ |___ /~~\\\\  |  |___    /~~\\\\ |    |    [39m
[32m                                                                                                       [39m
‚ÑπÔ∏è  –ó–∞–≥—Ä—É–∑–∫–∞ —à–∞–±–ª–æ–Ω–∞? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) [42D[42C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) t[43D[43C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) te[44D[44C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) tem[45D[45C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) temp[46D[46C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) templ[47D[47C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) templa[48D[48C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) templat[49D[49C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) template[50D[50C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) template-[51D[51C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) template-a[52D[52C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) template-ap[53D[53C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: (template-app) template-app[54D[54C[2K[G? –ù–∞–∑–≤–∞–Ω–∏–µ –≤–∞—à–µ–≥–æ –ø—Ä–æ–µ–∫—Ç–∞: template-app[39D[39C
? –í—ã–ø–æ–ª–Ω–∏—Ç—å —É—Å—Ç–∞–Ω–æ–≤–∫—É –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–µ–π (npm install)? (Y/n) [56D[56C[2K[G? –í—ã–ø–æ–ª–Ω–∏—Ç—å —É—Å—Ç–∞–Ω–æ–≤–∫—É –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–µ–π (npm install)? (Y/n) n[57D[57C[2K[G? –í—ã–ø–æ–ª–Ω–∏—Ç—å —É—Å—Ç–∞–Ω–æ–≤–∫—É –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–µ–π (npm install)? No[52D[52C

‚úÖ –ì–æ—Ç–æ–≤–æ!

"
`;
